ext/opcache/Optimizer/dfa_pass.c

```diff
@@ -895,48 +896,29 @@ static int zend_dfa_optimize_jmps(zend_op_array *op_array, zend_ssa *ssa)
 					break;
 				}
 				case ZEND_SWITCH_LONG:
-					if (opline->op1_type == IS_CONST) {
-						zval *zv = CT_CONSTANT_EX(op_array, opline->op1.constant);
-						if (Z_TYPE_P(zv) != IS_LONG) {
-							removed_ops++;
-							MAKE_NOP(opline);
-							opline->extended_value = 0;
-							take_successor_ex(ssa, block_num, block, block->successors[0]);
-							goto optimize_nop;
-						} else {
-							HashTable *jmptable = Z_ARRVAL_P(CT_CONSTANT_EX(op_array, opline->op2.constant));
-							zval *jmp_zv = zend_hash_index_find(jmptable, Z_LVAL_P(zv));
-							uint32_t target;
-
-							if (jmp_zv) {
-								target = ZEND_OFFSET_TO_OPLINE_NUM(op_array, opline, Z_LVAL_P(jmp_zv));
-							} else {
-								target = ZEND_OFFSET_TO_OPLINE_NUM(op_array, opline, opline->extended_value);
-							}
-							opline->opcode = ZEND_JMP;
-							opline->extended_value = 0;
-							SET_UNUSED(opline->op1);
-							ZEND_SET_OP_JMP_ADDR(opline, opline->op1, op_array->opcodes + target);
-							SET_UNUSED(opline->op2);
-							take_successor_ex(ssa, block_num, block, ssa->cfg.map[target]);
-							goto optimize_jmp;
-						}
-					}
-					break;
 				case ZEND_SWITCH_STRING:
+				case ZEND_MATCH:
 					if (opline->op1_type == IS_CONST) {
 						zval *zv = CT_CONSTANT_EX(op_array, opline->op1.constant);
-						if (Z_TYPE_P(zv) != IS_STRING) {
+						zend_uchar type = Z_TYPE_P(zv);
+						zend_bool correct_type =
+							(opline->opcode == ZEND_SWITCH_LONG && type == IS_LONG)
+							|| (opline->opcode == ZEND_SWITCH_STRING && type == IS_STRING)
+							|| (opline->opcode == ZEND_MATCH && (type == IS_LONG || type == IS_STRING));
+
+						if (!correct_type) {
 							removed_ops++;
 							MAKE_NOP(opline);
 							opline->extended_value = 0;
 							take_successor_ex(ssa, block_num, block, block->successors[0]);
```

Comment:
I'm confused by this ... how does this even work for switch? It takes successor 0, which corresponds to the first jumptable entry and doesn't really make sense.

Reply:
Okay, it looks like this is effectively dead code, because all cases that would hit this already get folded by SCCP. To avoid confusion, I'd still change this to use `block->successors[block->successor_count - 1]` (which should work both for switch (fallthrough to compare chain) and match (default destination).